examples/step-58/step-58.cc

```diff
@@ -0,0 +1,660 @@
+  // Complex numbers are not easily visualized. We can output their real and
+  // imaginary parts, i.e., the fields $\text{Re}(\psi_h^{(n)}(\mathbf x))$ and
+  // $\text{Im}(\psi_h^{(n)}(\mathbf x))$, and that is exactly what the DataOut
+  // class does when one attaches as complex-valued vector via
+  // DataOut::add_data_vector() and then calls DataOut::build_patches(). But
+  // oftentimes we are not actually interested in real and imaginary parts of
+  // the solution vector, but instead in derived quantities such as the
+  // magnitude and phase angle of the solution. The way to put computed
+  // quantities into output files for visualization -- as used in numerous
+  // previous tutorial programs -- is to use the facilities of the
+  // DataPostprocessor and derived classes. Specifically, both the magnitude of
+  // a complex number and its phase angles are scalar quantities, and so the
+  // DataPostprocessorScalar class is the right tool to base what we want to do
+  // on.
+  //
+  // Consequently, what we do here is to implement two classes
+  // `ComplexMagnitude` and `ComplexPhase` that compute for each point at which
+  // DataOut decides to generate output, the magnitudes $|\psi_h|$ and phases
+  // $\text{arg}(\psi_h)$ of the solution for visualization. There is a fair
+  // amount of boiler-plate code below, with the only interesting parts of
+  // the first of these two classes being how its `evaluate_vector_field()`
+  // function computes the `computed_quantities` object.
+  namespace DataPostprocessors
+  {
+    template <int dim>
+    class ComplexMagnitude : public DataPostprocessorScalar<dim>
+    {
+    public:
+      ComplexMagnitude();
+
+      virtual void evaluate_vector_field(
+        const DataPostprocessorInputs::Vector<dim> &inputs,
+        std::vector<Vector<double>> &computed_quantities) const override;
+    };
+
+
+    template <int dim>
+    ComplexMagnitude<dim>::ComplexMagnitude()
+      : DataPostprocessorScalar<dim>("Magnitude", update_values)
+    {}
+
+
+    template <int dim>
+    void ComplexMagnitude<dim>::evaluate_vector_field(
+      const DataPostprocessorInputs::Vector<dim> &inputs,
+      std::vector<Vector<double>> &               computed_quantities) const
+    {
+      Assert(computed_quantities.size() == inputs.solution_values.size(),
+             ExcDimensionMismatch(computed_quantities.size(),
+                                  inputs.solution_values.size()));
+
+      for (unsigned int q = 0; q < computed_quantities.size(); ++q)
+        {
+          Assert(computed_quantities[q].size() == 1,
+                 ExcDimensionMismatch(computed_quantities[q].size(), 1));
+          Assert(inputs.solution_values[q].size() == 2,
+                 ExcDimensionMismatch(inputs.solution_values[q].size(), 2));
+
+          computed_quantities[q](0) =
+            std::abs(std::complex<double>(inputs.solution_values[q](0),
+                                          inputs.solution_values[q](1)));
```

Comment:
Great example of `DataProcessor` for complex valued equations!

Maybe you can mention in the text that the number of inputs is equal to two times the number of components. In this case you have one component, therefore you have two inputs. I guess that if you would have 3 components. You would have 6 inputs. Right?

In the case of 3 components (vector field). Which would be the order of the inputs:
- x_re, y_re, z_re, x_im, y_im, z_im
- x_re, x_im, y_re, y_im, z_re, z_im

I think that it would be good to document this.

Reply:
That's actually discussed at length in `DataPostprocessor`. The order is `x_re, x_im, y_re, y_im, ...`.